tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c

```diff
@@ -0,0 +1,40 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_TCP_IP.h"
+
+#include "../../utility/memory_assignments.c"
+
+/* This proof assumes that prvTCPPrepareSend and prvTCPReturnPacket are correct.
+These functions are proved to be correct separately. */
+
+BaseType_t publicTCPHandleState( FreeRTOS_Socket_t *pxSocket, NetworkBufferDescriptor_t **ppxNetworkBuffer );
+
+void harness() {
+	FreeRTOS_Socket_t *pxSocket = ensure_FreeRTOS_Socket_t_is_allocated();
+	if (ensure_memory_is_valid(pxSocket)) {
+		pxSocket->u.xTCP.rxStream = ensure_FreeRTOS_StreamBuffer_t_is_allocated();
+		pxSocket->u.xTCP.txStream = ensure_FreeRTOS_StreamBuffer_t_is_allocated();
+		pxSocket->u.xTCP.pxPeerSocket = ensure_FreeRTOS_Socket_t_is_allocated();
+		__CPROVER_assume(pxSocket->u.xTCP.xTCPWindow.ucOptionLength == sizeof(uint32_t) * ipSIZE_TCP_OPTIONS/sizeof(uint32_t));
+		__CPROVER_assume(pxSocket->u.xTCP.uxRxWinSize >= 0 && pxSocket->u.xTCP.uxRxWinSize <= sizeof(size_t));
+		__CPROVER_assume(pxSocket->u.xTCP.usInitMSS == sizeof(uint16_t));
+	}
+
+	NetworkBufferDescriptor_t *pxNetworkBuffer = ensure_FreeRTOS_NetworkBuffer_is_allocated();
+	if(ensure_memory_is_valid(pxNetworkBuffer)) {
+		pxNetworkBuffer->pucEthernetBuffer = safeMalloc(sizeof(TCPPacket_t));
```

Comment:
What decides what goes into the ```ensure``` function, and what you do manually?```

Reply:
`ensure_memory_is_valid` macro checks if pointer is not `NULL`. The pointer can be pointing to anything. Before this I wrote manually`if(px != NULL)` to check for NULL pointers.

Reply:
Would it make sense to use ```__CPROVER_w_ok((base), (len))``` to check the memory was valid and not just non-null

Reply:
Sure. I have added it.